// linaria/lib/rollup.js
"use strict";

require("core-js/modules/es.string.replace");

exports.__esModule = true;
exports.default = linaria;

var _rollupPluginutils = require("rollup-pluginutils");

var _transform = _interopRequireDefault(require("./transform"));

var _slugify = _interopRequireDefault(require("./slugify"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * This file contains a Rollup loader for Linaria.
 * It uses the transform.ts function to generate class names from source code,
 * returns transformed code without template literals and attaches generated source maps
 */
function linaria({
  include,
  exclude,
  sourceMap,
  preprocessor,
  ...rest
} = {}) {
  const filter = (0, _rollupPluginutils.createFilter)(include, exclude);
  const cssLookup = {};
  return {
    name: 'linaria',

    load(id) {
      return cssLookup[id];
    },

    /* eslint-disable-next-line consistent-return */
    resolveId(importee) {
      if (importee in cssLookup) return importee;
    },

    transform(code, id) {
      // Do not transform ignored and generated files
      if (!filter(id) || id in cssLookup) return;
      const result = (0, _transform.default)(code, {
        filename: id,
        preprocessor,
        pluginOptions: rest
      });
      if (!result.cssText) return;
      let {
        cssText
      } = result;
      const slug = (0, _slugify.default)(cssText);
      const filename = `${id.replace(/\.js$/, '')}_${slug}.css`;

      if (sourceMap && result.cssSourceMapText) {
        const map = Buffer.from(result.cssSourceMapText).toString('base64');
        cssText += `/*# sourceMappingURL=data:application/json;base64,${map}*/`;
      }

      cssLookup[filename] = cssText;
      result.code += `\nimport ${JSON.stringify(filename)};\n`;
      /* eslint-disable-next-line consistent-return */

      return {
        code: result.code,
        map: result.sourceMap
      };
    }

  };
}
